signac/contrib/project.py

```diff
@@ -1526,11 +1518,45 @@ def init_project(cls, name, root=None, workspace=None, make_dir=True):
             configuration.
 
         """
+        # TODO: Remove both the `if args` and `if kwargs` blocks in version 3.0
+        # when we remove backwards compatibility for project name APIs.
+        if args:
+            num_args = len(args)
+            if num_args == 1:
+                warnings.warn(
+                    "Project names were removed in signac 2.0. If you intended to call "
+                    "`init_project` with a root directory, please provided it as a keyword "
+                    f"argument: `init_project(root={args[0]})`. If your project name "
+                    "contains important information, consider storing it in the project document "
+                    "instead.",
```

Comment:
Should we store it automatically similar to the proposed migration? I feel like simply dropping this piece of information (even with warning) is risky. I think if we actually just completely ignore this information, we need to fail instead.

Reply:
This PR removes the ability for a user to call the API with a project name, but it does not change the schema or remove the project name internally. If a user has already created a project with a name other than project (or if they initialize a project and then modify `signac.rc` directly), calling `signac.get_project` here will still provide the name they used. I did this in attempt to keep this PR's scope to a minimum and would prefer to leave it that way since there were questions like this one that I would prefer to address in a separate PR dedicated to logic for schema migration. Note that this PR is being merged into `next`, not `schema2`, again in an attempt to completely separate API changes from schema changes except where the two are inextricably coupled.

Reply:
In [this discussion](https://github.com/glotzerlab/signac/pull/643#discussion_r786069407) on #643 we decided that during a migration we would offer to automatically move the project name into the project document. In [this other discussion](https://github.com/glotzerlab/signac/pull/643#discussion_r752469873) we proposed changing the `init_project` API to warn users that the project name will be ignored when provided to the `init_project` API in signac 2.x. The goal would be to avoid destroying preexisting provenance while providing a smooth transition to the new API without breaking existing scripts. Am I interpreting these discussions correctly?

It seems like what you are proposing now is to additionally not just warn users, but to also to e.g. offer to automatically store the project name in the project document (one option for addressing the "mitigation issue" that you describe) if a user of signac 2.x calls `init_project(name=...)`. It's one thing for the migration to offer to preserve preexisting names, but should we really provide that feature on new project creation in signac 2.x? I would rather make our warning more explicit: "If the project name is important to you, consider storing it in the project document." We could offer to store project names in `init_project` as well if you think that it's particularly important, but I suspect that will encourage more users to avoid updating their usage of the API until 3.x. We would also need to continue maintaining machinery to check that the provided project name matches one in the document if it exists and then decide how to act in the case of a mismatch since it's no longer critical provenance (error? warn? overwrite? ignore?).

Reply:
I understand that some of the decisions we made in those past discussions are slightly contradictory. I simply did not sufficiently consider that the proposed approach with a warning is good to not break existing code, but it would also mean to dump user-provided metadata into `/dev/null`.

I would propose to _either_:

1. Keep the warning, but also store the project name in the project document under a "name" key. The warning should explicitly say that this is what happened. This should fail if the key already exists with a different value.
2. Raise an error and advise users to store the project name in the project document if they consider it important.

In principle, we should never destroy user-provided information IMO.

Reply:
No worries, I didn't mean to imply any issue with changing direction. I just wanted to make sure that we were on the same page as to what has been discussed since at this point discussions are fragmented over multiple PRs and over a reasonably long period of time.

I think option 1 is probably what we should do. My personal preference would be option 2 since it is simpler for us, but realistically I think converting `init_project(name=...)` into an error will break enough scripts that we're better off at least giving people the opportunity to move off more gradually. In practice I think a large number of users will just ignore the warning anyway, so we'll just run into the same issues when people upgrade to 3.x, but it's better than nothing.